modules/core/src/count_non_zero.simd.hpp

```diff
@@ -179,7 +179,25 @@ static int countNonZero32f( const float* src, int len )
 
 static int countNonZero64f( const double* src, int len )
 {
-    return countNonZero_(src, len);
+    int nz = 0, i = 0;
+#if CV_SIMD_64F
+    v_int64 sum1 = vx_setzero_s64();
+    v_int64 sum2 = vx_setzero_s64();
+    v_float64 zero = vx_setzero_f64();
+    int step = v_float64::nlanes * 2;
+    int len0 = len & -step;
+
+    for(i = 0; i < len0; i += step )
+        {
+        sum1 += v_reinterpret_as_s64(vx_load(&src[i]) == zero);
```

Comment:
Can't these loads and comparisons be done with u32 instead?  Comparing to zero is the same and should be faster.

Reply:
Floating point numbers are tricky.
There are two zeros possible: +0, -0, so interpretation as int64/uint64 is not valid.
Details: https://en.wikipedia.org/wiki/Signed_zero

BTW, The same bug is in `countNonZero32f()` optimization.

Reply:
How about using a 63-bit mask on top of that?

v_zero_mask = 0x7fffffffffffffff;

((vx_load(&src[i]) & v_zero_mask) == zero)

Reply:
Right!
Mask should work (on known platforms).

Reply:
This is a subtle IEEE nuance. Assuming a compliant IEEE754-2008 implementation, the sign of zero is ignored. (chapter 5.11)

Reply:
Yes, floating-point comparison works good.
but it is not equal to integer-based zero check:

> comparisons be done with s64 instead

f32 optimization is fine (v_reinterpret applied after comparison).
Check in current implementation is fine too (due floating-point check).